README.md

# VariAnalyser

<!-- badges: start -->

<!-- badges: end -->

The goal of VariAnalyser is to functionally annotate list of variants from 
Genome-wide Association Studies (GWAS). Particularly, VariAnalyser considers 
CADD functional predictions, Gene Oncology (GO) and KEGG pathways and disease 
associations with genes.

## Installation

You can install the most recent version of VariAnalyser like so:

``` r
if(!require("devtools")){
    install.packages("devtools")
    library("devtools")
}
install_github("Broccolito/VariAnalyser")
```

## Example

To annotate the GWAS results and include pathway and disease information

``` r
library(VariAnalyser)
PathwayDiseaseAnnotator("FHS_EA_MRS_merged.txt")
```

To make a Manhattan plot

```r
ManhattanPlotter("FHS_EA_MRS_merged.txt")
```



NEWS.md
# VariAnalyser 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/ManhattanPlotter.R
#' Plot a Manhattan plot and save it as a png file
#'
#' @param gwas_result_file The name of the text file containing GWAS statistics
#' @param chr The name of the chromosome number column in the text file
#' @param pos The name of the chromosome position column in the text file
#' @param ref The name of the reference allele column in the text file
#' @param alt The name of the alternative allele column in the text file
#' @param pvalue The name of the pvalue column in the text file
#' @param delimiter The delimiter used in the text file
#' @param color1 First color of the Manhattan plot
#' @param color2 Second color of the Manhattan plot
#' @param threshold_linecolor Default is set as black
#'
#' @return A ggplot object of the Manhattan plot generated
#'
#' @examples
#' ManhattanPlotter("data/example/FHS_EA_MRS_merged.txt")

ManhattanPlotter = function(gwas_result_file = "data/example/FHS_EA_MRS_merged.txt",
                            chr = "CHR",
                            pos = "POS",
                            pvalue = "p.value",
                            delimiter = " ",
                            color1 = "cornflowerblue",
                            color2 = "blue4",
                            threshold_linecolor = "black"){

  cat("Loading GWAS Statistics...\n")
  data = read.delim(gwas_result_file, sep = delimiter)
  cat("GWAS Statistics Loaded...\n")

  names(data)[which(names(data)==chr)] = "chr"
  names(data)[which(names(data)==pos)] = "pos"
  names(data)[which(names(data)==pvalue)] = "pvalue"

  data = data %>%
    select(chr, pos, pvalue) %>%
    mutate(chr = ifelse(chr == "X", "23", chr)) %>%
    mutate(chr = as.numeric(chr)) %>%
    mutate(pos = as.numeric(pos)) %>%
    mutate(pvalue = as.numeric(pvalue))

  data_chromosome_position = data %>%
    group_by(chr) %>%
    summarise(chromosome_length = max(pos)) %>%
    mutate(tot = cumsum(as.numeric(chromosome_length)) - chromosome_length) %>%
    select(-chromosome_length)

  data_cleaned = left_join(data, data_chromosome_position, by = "chr") %>%
    arrange(chr, pos) %>%
    mutate(bp_cum = pos + tot) %>%
    mutate(nlogp = -log(pvalue, base = 10)) %>%
    select(chr, pos, pvalue, nlogp, tot, bp_cum)

  axisdf = data_cleaned %>%
    group_by(chr) %>% summarize(center = (max(bp_cum) + min(bp_cum))/2)

  plt = ggplot(data_cleaned, aes(x = bp_cum, y = nlogp)) +
    geom_point(aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c(color1, color2), 22)) +
    scale_x_continuous(label = axisdf$chr, breaks = axisdf$center) +
    scale_y_continuous(limits = c(0, max(max(data_cleaned$nlogp)+1,
                                         -log(5e-8,base = 10)+1)),
                       expand = c(0, 0)) +
    geom_hline(yintercept = -log(5e-8,base = 10),
               linetype = "solid", color = threshold_linecolor) +
    geom_hline(yintercept = -log(1e-5,base = 10),
               linetype = "dashed", color = threshold_linecolor) +
    xlab("Chromosome") +
    ylab(TeX("-log_{10}(p-value)")) +
    theme_pubr() +
    theme(
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.title.y = element_text(size = 20),
      axis.title.x = element_text(size = 20)
    )

  cat("Saving the Manhattan Plot as a PNG file...\n")
  plot_name = gsub(pattern = ".txt", replacement = "_ManhattanPlot.png", gwas_result_file)
  ggsave(filename = plot_name, plot = plt, device = "png", dpi = 1200,
         height = 6, width = 15)
  cat(paste0("Plot saved as ", plot_name, "\n"))

  return(plt)

}



R/VariantAnnotator.R
#' Annotate the GWAS results
#'
#' @param variant_file The name of the text file containing GWAS statistics
#' @param chr The name of the chromosome number column in the text file
#' @param pos The name of the chromosome position column in the text file
#' @param ref The name of the reference allele column in the text file
#' @param alt The name of the alternative allele column in the text file
#' @param pvalue The name of the pvalue column in the text file
#' @param delimiter The delimiter used in the text file
#' @param write_csv Whether to write the result to a csv file, else skip saving
#'
#' @return A data frame of the annotated results
#'
#' @examples
#' VariantAnnotator("data/example/FHS_EA_MRS_merged.txt")


VariantAnnotator = function(variant_file = "data/example/FHS_EA_MRS_5e8_snplist.txt",
                            chr = "CHR",
                            pos = "POS",
                            ref = "NEA",
                            alt = "EA",
                            delimiter = " ",
                            write_csv = TRUE){

  build_convert = function(chromosome = 15,
                           pos = 48481729){
    grObject = GRanges(seqnames = paste0("chr", chromosome),
                       ranges = IRanges(start=pos, end=pos))
    results = as.data.frame(liftOver(grObject, chainObject_tohg38))
    results = paste0(results$seqnames, ":", results$start)
    return(results)
  }

  cat("Loading Variant File...\n")
  data = read.delim(variant_file, sep = delimiter)

  names(data)[which(names(data)==chr)] = "chr"
  names(data)[which(names(data)==pos)] = "pos"
  names(data)[which(names(data)==ref)] = "ref"
  names(data)[which(names(data)==alt)] = "alt"

  data = data %>%
    mutate(chr = ifelse(chr == "X", "23", chr)) %>%
    mutate(chr = as.numeric(chr)) %>%
    mutate(pos = as.numeric(pos))

  data = data %>%
    mutate(query = paste0("chr", chr, ":g.", pos, ref, ">", alt))

  cat("Converting Variant Position to hg38...\n")
  data$hg38build = mapply(build_convert, data$chr, data$pos)

  cat("Generating Variant Annotation...\n")
  variant_client = BioThingsClient("variant")
  annotation = vector()
  for(q in data$query){
    try({
      v = btGet(variant_client, q)[[1]]

      gene_symbol = NA
      gene_name = NA
      variant_type = NA
      variant_consequence = NA
      variant_CADD_phred = NA
      variant_phylop_vertebrate = NA
      variant_polyphen = NA
      variant_sift = NA
      variant_rs = NA

      gene_symbol = paste(v[["dbsnp"]][["gene"]][["symbol"]], collapse = "; ")
      gene_name = gsub(pattern = ",", ";", v[["dbsnp"]][["gene"]][["name"]])
      variant_type = gsub(",","; ", paste(v[["cadd"]][["consdetail"]],
                                          collapse = "; "))
      variant_consequence = gsub(",","; ", paste(v[["cadd"]][["consequence"]],
                                                 collapse = "; "))
      variant_CADD_phred = v[["cadd"]][["phred"]]
      variant_phylop_vertebrate = v[["cadd"]][["phylop"]][["vertebrate"]]
      variant_polyphen = v[["cadd"]][["polyphen"]][["val"]]
      variant_sift = v[["cadd"]][["sift"]][["val"]]
      variant_rs = v[["dbsnp"]][["rsid"]]

      gene_symbol = ifelse(is.null(gene_symbol), NA, gene_symbol)
      gene_name = ifelse(is.null(gene_name), NA, gene_name)
      variant_type = ifelse(is.null(variant_type), NA, variant_type)
      variant_consequence = ifelse(is.null(variant_consequence), NA, variant_consequence)

      variant_CADD_phred = ifelse(is.null(variant_CADD_phred), NA, variant_CADD_phred)
      variant_phylop_vertebrate = ifelse(is.null(variant_phylop_vertebrate),
                                         NA, variant_phylop_vertebrate)
      variant_polyphen = ifelse(is.null(variant_polyphen), NA, variant_polyphen)
      variant_sift = ifelse(is.null(variant_sift), NA, variant_sift)
      variant_rs = ifelse(is.null(variant_rs), NA, variant_rs)

      anno = tibble(query = q, variant_rs, gene_symbol, gene_name, variant_type,
                    variant_consequence, variant_CADD_phred,
                    variant_phylop_vertebrate, variant_polyphen,
                    variant_sift)

      annotation = rbind.data.frame(annotation, anno)
      cat(paste0("Annotating ", q," ..\n"))
    }, silent = TRUE)
  }

  data_annotated = data %>%
    left_join(annotation, by = "query")

  if(write_csv){
    file_name = gsub(pattern = ".txt",
                     replacement = "_Annotated.csv",
                     variant_file)
    write.csv(data_annotated, file = file_name,
              quote = FALSE, row.names = FALSE)
  }

  return(data_annotated)

}


R/PathwayDiseaseAnnotator.R
#' Annotate the GWAS results and include pathway and disease information
#'
#' @param variant_file The name of the text file containing GWAS statistics
#' @param chr The name of the chromosome number column in the text file
#' @param pos The name of the chromosome position column in the text file
#' @param ref The name of the reference allele column in the text file
#' @param alt The name of the alternative allele column in the text file
#' @param pvalue The name of the pvalue column in the text file
#' @param delimiter The delimiter used in the text file
#' @param write_excel Whether to write the result to an excel file, else write as CSV
#'
#' @return A data frame of the annotated results
#'
#' @examples
#' PathwayDiseaseAnnotator("data/example/FHS_EA_MRS_merged.txt")


PathwayDiseaseAnnotator = function(variant_file = "data/example/FHS_EA_MRS_5e8_snplist.txt",
                                   chr = "CHR",
                                   pos = "POS",
                                   ref = "NEA",
                                   alt = "EA",
                                   delimiter = " ",
                                   write_excel = TRUE){

  data("genes")
  data("diseases_pathways_flatten")
  data("gene_pathways_flatten")

  cat("Annotating variants...\n")
  variants_annotated = VariantAnnotator(variant_file,
                                        chr,
                                        pos,
                                        ref,
                                        alt,
                                        delimiter,
                                        write_csv = FALSE)

  cat("Annotating Diseases and Pathways...\n")
  data_annotated = variants_annotated %>%
    mutate(GeneSymbol = gene_symbol) %>%
    left_join(select(genes, GeneSymbol, GeneID),
              by = "GeneSymbol") %>%
    left_join(gene_pathways_flatten, "GeneSymbol") %>%
    left_join(diseases_pathways_flatten, by = "GeneSymbol") %>%
    select(-GeneSymbol)

  cat("Saving Files...\n")
  if(write_excel){
    cat("Saving as an excel file...\n")
    file_name = gsub(pattern = ".txt",
                     replacement = "_PathwayDiseaseAnnotated.xlsx",
                     variant_file)
    write_xlsx(data_annotated, path = file_name)
  }else{
    cat("Saving as a csv file...\n")
    file_name = gsub(pattern = ".txt",
                     replacement = "_PathwayDiseaseAnnotated.csv",
                     variant_file)
    write.csv(data_annotated, file = file_name,
              quote = FALSE, row.names = FALSE)
  }

  return(data_annotated)

}












R/InteractiveManhattanPlotter.R
#' Plot an interactive Manhattan plot and save it as an HTML
#'
#' @param gwas_result_file The name of the text file containing GWAS statistics
#' @param chr The name of the chromosome number column in the text file
#' @param pos The name of the chromosome position column in the text file
#' @param ref The name of the reference allele column in the text file
#' @param alt The name of the alternative allele column in the text file
#' @param pvalue The name of the pvalue column in the text file
#' @param delimiter The delimiter used in the text file
#' @param color1 First color of the Manhattan plot
#' @param color2 Second color of the Manhattan plot
#' @param threshold_linecolor Default is set as black
#'
#' @return A html widget of the Manhattan plot generated
#'
#' @examples
#' InteractiveManhattanPlotter("data/example/FHS_EA_MRS_merged.txt")


InteractiveManhattanPlotter = function(gwas_result_file = "data/example/FHS_EA_MRS_merged.txt",
                                       chr = "CHR",
                                       pos = "POS",
                                       ref = "NEA",
                                       alt = "EA",
                                       pvalue = "p.value",
                                       delimiter = " ",
                                       color1 = "cornflowerblue",
                                       color2 = "blue4",
                                       threshold_linecolor = "black"){

  cat("Loading GWAS Statistics...\n")
  data = read.delim(gwas_result_file, sep = delimiter)
  cat("GWAS Statistics Loaded...\n")

  names(data)[which(names(data)==chr)] = "chr"
  names(data)[which(names(data)==pos)] = "pos"
  names(data)[which(names(data)==ref)] = "ref"
  names(data)[which(names(data)==alt)] = "alt"
  names(data)[which(names(data)==pvalue)] = "pvalue"

  data = data %>%
    select(chr, pos, ref, alt, pvalue) %>%
    mutate(chr = ifelse(chr == "X", "23", chr)) %>%
    mutate(chr = as.numeric(chr)) %>%
    mutate(pos = as.numeric(pos)) %>%
    mutate(pvalue = as.numeric(pvalue))

  data_chromosome_position = data %>%
    group_by(chr) %>%
    summarise(chromosome_length = max(pos)) %>%
    mutate(tot = cumsum(as.numeric(chromosome_length)) - chromosome_length) %>%
    select(-chromosome_length)

  data_cleaned = left_join(data, data_chromosome_position, by = "chr") %>%
    arrange(chr, pos) %>%
    mutate(bp_cum = pos + tot) %>%
    mutate(nlogp = -log(pvalue, base = 10)) %>%
    select(chr, pos, ref, alt, pvalue, nlogp, tot, bp_cum)

  axisdf = data_cleaned %>%
    group_by(chr) %>% summarize(center = (max(bp_cum) + min(bp_cum))/2)

  data_cleaned = data_cleaned %>%
    filter(nlogp >= 5) %>%
    mutate(query = paste0("chr", chr, ":g.", pos, ref, ">", alt))

  if(dim(data_cleaned)[1]<1){
    return("No Significant Signal Detected...\n")
  }

  cat("Generating Variant Annotation...\n")
  variant_client = BioThingsClient("variant")
  annotation = vector()
  for(q in data_cleaned$query){
    try({
      v = btGet(variant_client, q)[[1]]

      gene_symbol = NA
      gene_name = NA
      variant_type = NA
      variant_consequence = NA
      variant_CADD_phred = NA
      variant_phylop_vertebrate = NA
      variant_polyphen = NA
      variant_sift = NA
      variant_rs = NA

      gene_symbol = paste(v[["dbsnp"]][["gene"]][["symbol"]], collapse = "; ")
      gene_name = gsub(pattern = ",", ";", v[["dbsnp"]][["gene"]][["name"]])
      variant_type = gsub(",","; ", paste(v[["cadd"]][["consdetail"]],
                                          collapse = "; "))
      variant_consequence = gsub(",","; ", paste(v[["cadd"]][["consequence"]],
                                                 collapse = "; "))
      variant_CADD_phred = v[["cadd"]][["phred"]]
      variant_phylop_vertebrate = v[["cadd"]][["phylop"]][["vertebrate"]]
      variant_polyphen = v[["cadd"]][["polyphen"]][["val"]]
      variant_sift = v[["cadd"]][["sift"]][["val"]]
      variant_rs = v[["dbsnp"]][["rsid"]]

      gene_symbol = ifelse(is.null(gene_symbol), NA, gene_symbol)
      gene_name = ifelse(is.null(gene_name), NA, gene_name)
      variant_type = ifelse(is.null(variant_type), NA, variant_type)
      variant_consequence = ifelse(is.null(variant_consequence), NA, variant_consequence)

      variant_CADD_phred = ifelse(is.null(variant_CADD_phred), NA, variant_CADD_phred)
      variant_phylop_vertebrate = ifelse(is.null(variant_phylop_vertebrate),
                                         NA, variant_phylop_vertebrate)
      variant_polyphen = ifelse(is.null(variant_polyphen), NA, variant_polyphen)
      variant_sift = ifelse(is.null(variant_sift), NA, variant_sift)
      variant_rs = ifelse(is.null(variant_rs), NA, variant_rs)

      anno = tibble(query = q, gene_symbol, gene_name, variant_type,
                    variant_consequence, variant_CADD_phred,
                    variant_phylop_vertebrate, variant_polyphen,
                    variant_sift, variant_rs)

      annotation = rbind.data.frame(annotation, anno)
      cat(paste0("Annotating ", q," ..\n"))
    }, silent = TRUE)
  }

  data_cleaned = data_cleaned %>%
    left_join(annotation, by = "query")

  cat("Generating Interactive Manhattan Plot of significant Signals...\n")
  suppressWarnings({
    plt = ggplot(data_cleaned, aes(x = bp_cum, y = nlogp)) +
      geom_point(aes(color=as.factor(chr),
                     query = query,
                     variant_rs = variant_rs,
                     gene_symbol = gene_symbol,
                     gene_name = gene_name,
                     variant_type = variant_type,
                     variant_consequence = variant_consequence,
                     variant_CADD_phred = variant_CADD_phred,
                     variant_phylop_vertebrate = variant_phylop_vertebrate,
                     variant_polyphen = variant_polyphen,
                     variant_sift = variant_sift),
                 alpha=0.8, size=1.3) +
      scale_color_manual(values = rep(c(color1, color2), 22)) +
      scale_x_continuous(label = axisdf$chr, breaks = axisdf$center) +
      scale_y_continuous(limits = c(4, max(max(data_cleaned$nlogp)+1,
                                           -log(5e-8,base = 10)+1)),
                         expand = c(0, 0)) +
      geom_hline(yintercept = -log(5e-8,base = 10),
                 linetype = "solid", color = threshold_linecolor) +
      geom_hline(yintercept = -log(1e-5,base = 10),
                 linetype = "dashed", color = threshold_linecolor) +
      xlab("Chromosome") +
      ylab("-log(p-value)") +
      theme_pubr() +
      theme(
        legend.position="none",
        panel.border = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        axis.title.y = element_text(size = 20),
        axis.title.x = element_text(size = 20)
      )
  })

  plt = ggplotly(plt, tooltip = c("nlogp", "query", "variant_rs",
                                  "gene_symbol", "gene_name", "variant_type",
                                  "variant_consequence", "variant_CADD_phred",
                                  "variant_phylop_vertebrate", "variant_polyphen",
                                  "variant_sift"))

  plot_name = gsub(pattern = ".txt",
                   replacement = "_InteractiveManhattanPlot.html",
                   gwas_result_file)
  saveWidget(plt, plot_name, selfcontained = TRUE, libdir = "lib")

  return(plt)

}

